File: test/cpp/test_forest.cpp
/*!
 * Derived from xgboost tree unit test code:
 * https://github.com/dmlc/xgboost/blob/master/tests/cpp/tree/test_tree_model.cc
 */
#include <gtest/gtest.h>
#include <testutils.h>
#include <stochtree/data.h>
#include <stochtree/ensemble.h>
#include <stochtree/log.h>
#include <stochtree/tree.h>
#include <iostream>
#include <memory>

TEST(Forest, UnivariateForestConstruction) {
  int num_trees = 2;
  int output_dimension = 1;
  bool is_leaf_constant = true;
  StochTree::TreeEnsemble forest(num_trees, output_dimension, is_leaf_constant);
  StochTree::Tree* tree_1_ptr = forest.GetTree(0);
  StochTree::Tree* tree_2_ptr = forest.GetTree(1);
  ASSERT_EQ(tree_1_ptr->LeafValue(0), 0.);
  tree_1_ptr->ExpandNode(0, 0, 0., 0., 0.);
  ASSERT_EQ(tree_1_ptr->NumNodes(), 3);
  ASSERT_EQ(tree_1_ptr->NodeType(0), StochTree::TreeNodeType::kNumericalSplitNode);
  tree_1_ptr->CollapseToLeaf(0, 0.);
  ASSERT_EQ(tree_1_ptr->NumValidNodes(), 1);
  tree_1_ptr->ExpandNode(0, 0, 0., 0., 0.);
  ASSERT_EQ(tree_1_ptr->NumValidNodes(), 3);
  ASSERT_EQ(tree_1_ptr->NodeType(0), StochTree::TreeNodeType::kNumericalSplitNode);
  ASSERT_EQ(tree_1_ptr->NumLeaves(), 2);
  ASSERT_FALSE(tree_1_ptr->IsLeaf(0));
  ASSERT_TRUE(tree_1_ptr->IsLeaf(1));
  ASSERT_TRUE(tree_1_ptr->IsLeaf(2));
}

TEST(Forest, UnivariateForestMerge) {
  // Load test data
  StochTree::TestUtils::TestDataset test_dataset;
  test_dataset = StochTree::TestUtils::LoadSmallDatasetUnivariateBasis();
  std::vector<StochTree::FeatureType> feature_types(test_dataset.x_cols, StochTree::FeatureType::kNumeric);

  // Construct datasets
  int n = test_dataset.n;
  StochTree::ForestDataset dataset = StochTree::ForestDataset();
  dataset.AddCovariates(test_dataset.covariates.data(), n, test_dataset.x_cols, test_dataset.row_major);
  dataset.AddBasis(test_dataset.omega.data(), test_dataset.n, test_dataset.omega_cols, test_dataset.row_major);
  StochTree::ColumnVector residual = StochTree::ColumnVector(test_dataset.outcome.data(), n);
  
  // Create a small ensemble
  int output_dim = 1;
  int num_trees = 2;
  bool is_leaf_constant = true;
  StochTree::TreeEnsemble ensemble1(num_trees, output_dim, is_leaf_constant);
  
  // Create another small ensemble
  StochTree::TreeEnsemble ensemble2(num_trees, output_dim, is_leaf_constant);

  // Simple split rules on both trees of forest 1
  auto* tree = ensemble1.GetTree(0);
  StochTree::TreeSplit tree_split = StochTree::TreeSplit(0.5);
  tree->ExpandNode(0, 0, tree_split, -5., 5.);
  tree = ensemble1.GetTree(1);
  tree->ExpandNode(0, 1, tree_split, -2.5, 2.5);

  // Run predict on the supplied covariates and check the result for the first forest
  std::vector<double> result(n*output_dim);
  std::vector<double> expected_pred = {7.5,2.5,-7.5,7.5,7.5,7.5,2.5,7.5,-7.5,-2.5};
  ensemble1.PredictInplace(dataset.GetCovariates(), result, 0);
  for (int i = 0; i < n; i++) {
    ASSERT_NEAR(expected_pred[i], result[i], 0.01);
  }

  // Simple split rules on both trees of forest 2
  tree = ensemble2.GetTree(0);
  tree_split = StochTree::TreeSplit(0.6);
  tree->ExpandNode(0, 1, tree_split, -1., 1.);
  tree = ensemble2.GetTree(1);
  tree_split = StochTree::TreeSplit(0.5);
  tree->ExpandNode(0, 2, tree_split, -0.5, 0.5);

  // Run predict on the supplied covariates and check the result for the second forest
  result = std::vector<double>(n*output_dim);
  expected_pred = std::vector<double>{1.5,-1.5,-0.5,-0.5,1.5,1.5,-1.5,1.5,-0.5,-1.5};
  ensemble2.PredictInplace(dataset.GetCovariates(), result, 0);
  for (int i = 0; i < n; i++) {
    ASSERT_NEAR(expected_pred[i], result[i], 0.01);
  }

  // Merge the second forest into the first
  ensemble1.MergeForest(ensemble2);
  result = std::vector<double>(n*output_dim);
  expected_pred = std::vector<double>{9,1,-8,7,9,9,1,9,-8,-4};
  ensemble1.PredictInplace(dataset.GetCovariates(), result, 0);
  for (int i = 0; i < n; i++) {
    ASSERT_NEAR(expected_pred[i], result[i], 0.01);
  }
}

TEST(Forest, UnivariateForestAdd) {
  // Load test data
  StochTree::TestUtils::TestDataset test_dataset;
  test_dataset = StochTree::TestUtils::LoadSmallDatasetUnivariateBasis();
  std::vector<StochTree::FeatureType> feature_types(test_dataset.x_cols, StochTree::FeatureType::kNumeric);

  // Construct datasets
  int n = test_dataset.n;
  StochTree::ForestDataset dataset = StochTree::ForestDataset();
  dataset.AddCovariates(test_dataset.covariates.data(), n, test_dataset.x_cols, test_dataset.row_major);
  dataset.AddBasis(test_dataset.omega.data(), test_dataset.n, test_dataset.omega_cols, test_dataset.row_major);
  StochTree::ColumnVector residual = StochTree::ColumnVector(test_dataset.outcome.data(), n);
  
  // Create a small ensemble
  int output_dim = 1;
  int num_trees = 2;
  bool is_leaf_constant = true;
  StochTree::TreeEnsemble ensemble1(num_trees, output_dim, is_leaf_constant);
  
  // Create another small ensemble
  StochTree::TreeEnsemble ensemble2(num_trees, output_dim, is_leaf_constant);

  // Simple split rules on both trees of forest 1
  auto* tree = ensemble1.GetTree(0);
  StochTree::TreeSplit tree_split = StochTree::TreeSplit(0.5);
  tree->ExpandNode(0, 0, tree_split, -5., 5.);
  tree = ensemble1.GetTree(1);
  tree->ExpandNode(0, 1, tree_split, -2.5, 2.5);

  // Run predict on the supplied covariates and check the result for the first forest
  std::vector<double> result(n*output_dim);
  std::vector<double> expected_pred = {7.5,2.5,-7.5,7.5,7.5,7.5,2.5,7.5,-7.5,-2.5};
  ensemble1.PredictInplace(dataset.GetCovariates(), result, 0);
  for (int i = 0; i < n; i++) {
    ASSERT_NEAR(expected_pred[i], result[i], 0.01);
  }

  // Add 1 to every leaf of the first forest
  ensemble1.AddValueToLeaves(1.0);

  // Run predict on the supplied covariates and check the result for the first forest
  result = std::vector<double>(n*output_dim);
  expected_pred = std::vector<double>{9.5,4.5,-5.5,9.5,9.5,9.5,4.5,9.5,-5.5,-0.5};
  ensemble1.PredictInplace(dataset.GetCovariates(), result, 0);
  for (int i = 0; i < n; i++) {
    ASSERT_NEAR(expected_pred[i], result[i], 0.01);
  }

  // Simple split rules on both trees of forest 2
  tree = ensemble2.GetTree(0);
  tree_split = StochTree::TreeSplit(0.6);
  tree->ExpandNode(0, 1, tree_split, -1., 1.);
  tree = ensemble2.GetTree(1);
  tree_split = StochTree::TreeSplit(0.5);
  tree->ExpandNode(0, 2, tree_split, -0.5, 0.5);

  // Run predict on the supplied covariates and check the result for the second forest
  result = std::vector<double>(n*output_dim);
  expected_pred = std::vector<double>{1.5,-1.5,-0.5,-0.5,1.5,1.5,-1.5,1.5,-0.5,-1.5};
  ensemble2.PredictInplace(dataset.GetCovariates(), result, 0);
  for (int i = 0; i < n; i++) {
    ASSERT_NEAR(expected_pred[i], result[i], 0.01);
  }

  // Subtract 1 from every leaf of the first forest
  ensemble2.AddValueToLeaves(-1.0);

  // Run predict on the supplied covariates and check the result for the first forest
  result = std::vector<double>(n*output_dim);
  expected_pred = std::vector<double>{-0.5,-3.5,-2.5,-2.5,-0.5,-0.5,-3.5,-0.5,-2.5,-3.5};
  ensemble2.PredictInplace(dataset.GetCovariates(), result, 0);
  for (int i = 0; i < n; i++) {
    ASSERT_NEAR(expected_pred[i], result[i], 0.01);
  }

  // Merge the second forest into the first
  ensemble1.MergeForest(ensemble2);
  result = std::vector<double>(n*output_dim);
  expected_pred = std::vector<double>{9,1,-8,7,9,9,1,9,-8,-4};
  ensemble1.PredictInplace(dataset.GetCovariates(), result, 0);
  for (int i = 0; i < n; i++) {
    ASSERT_NEAR(expected_pred[i], result[i], 0.01);
  }

  // Merge the second forest into the first
  ensemble1.MultiplyLeavesByValue(2.0);
  result = std::vector<double>(n*output_dim);
  expected_pred = std::vector<double>{18,2,-16,14,18,18,2,18,-16,-8};
  ensemble1.PredictInplace(dataset.GetCovariates(), result, 0);
  for (int i = 0; i < n; i++) {
    ASSERT_NEAR(expected_pred[i], result[i], 0.01);
  }
}

File: test/python/test_forest.py
import numpy as np

from stochtree import Dataset, Forest


class TestPredict:
    def test_constant_forest_construction(self):
        # Create dataset
        X = np.array([
            [1.5, 8.7, 1.2],
            [2.7, 3.4, 5.4],
            [3.6, 1.2, 9.3],
            [4.4, 5.4, 10.4],
            [5.3, 9.3, 3.6],
            [6.1, 10.4, 4.4],
        ])
        n, p = X.shape
        num_trees = 10
        output_dim = 1
        forest_dataset = Dataset()
        forest_dataset.add_covariates(X)
        forest = Forest(num_trees, output_dim, True, False)

        # Initialize a forest with constant root predictions
        forest.set_root_leaves(0.0)

        # Check that regular and "raw" predictions are the same (since the leaf is constant)
        pred = forest.predict(forest_dataset)
        pred_exp = np.array([0.0, 0.0, 0.0, 0.0, 0.0, 0.0])

        # Assertion
        np.testing.assert_almost_equal(pred, pred_exp)

        # Split the root of the first tree in the ensemble at X[,1] > 4.0
        forest.add_numeric_split(0, 0, 0, 4.0, -5.0, 5.0)

        # Check that regular and "raw" predictions are the same (since the leaf is constant)
        pred = forest.predict(forest_dataset)
        pred_exp = np.array([-5.0, -5.0, -5.0, 5.0, 5.0, 5.0])

        # Assertion
        np.testing.assert_almost_equal(pred, pred_exp)

        # Split the left leaf of the first tree in the ensemble at X[,2] > 4.0
        forest.add_numeric_split(0, 1, 1, 4.0, -7.5, -2.5)

        # Check that regular and "raw" predictions are the same (since the leaf is constant)
        pred = forest.predict(forest_dataset)
        pred_exp = np.array([-2.5, -7.5, -7.5, 5.0, 5.0, 5.0])

        # Assertion
        np.testing.assert_almost_equal(pred, pred_exp)

    def test_constant_forest_merge_arithmetic(self):
        # Create dataset
        X = np.array([
            [1.5, 8.7, 1.2],
            [2.7, 3.4, 5.4],
            [3.6, 1.2, 9.3],
            [4.4, 5.4, 10.4],
            [5.3, 9.3, 3.6],
            [6.1, 10.4, 4.4],
        ])
        n, p = X.shape
        num_trees = 10
        output_dim = 1
        forest_dataset = Dataset()
        forest_dataset.add_covariates(X)

        # Create two forests
        forest1 = Forest(num_trees, output_dim, True, False)
        forest2 = Forest(num_trees, output_dim, True, False)

        # Initialize forests with constant root predictions
        forest1.set_root_leaves(0.0)
        forest2.set_root_leaves(0.0)

        # Check that predictions are as expected
        pred1 = forest1.predict(forest_dataset)
        pred2 = forest2.predict(forest_dataset)
        pred_exp1 = np.array([0.0, 0.0, 0.0, 0.0, 0.0, 0.0])
        pred_exp2 = np.array([0.0, 0.0, 0.0, 0.0, 0.0, 0.0])

        # Assertion
        np.testing.assert_almost_equal(pred1, pred_exp1)
        np.testing.assert_almost_equal(pred2, pred_exp2)

        # Split the root of the first tree in the first forest at X[,1] > 4.0
        forest1.add_numeric_split(0, 0, 0, 4.0, -5.0, 5.0)

        # Split the root of the first tree in the second forest at X[,1] > 3.0
        forest2.add_numeric_split(0, 0, 0, 3.0, -1.0, 1.0)

        # Check that predictions are as expected
        pred1 = forest1.predict(forest_dataset)
        pred2 = forest2.predict(forest_dataset)
        pred_exp1 = np.array([-5.0, -5.0, -5.0, 5.0, 5.0, 5.0])
        pred_exp2 = np.array([-1.0, -1.0, 1.0, 1.0, 1.0, 1.0])

        # Assertion
        np.testing.assert_almost_equal(pred1, pred_exp1)
        np.testing.assert_almost_equal(pred2, pred_exp2)

        # Split the left leaf of the first tree in the first forest at X[,2] > 4.0
        forest1.add_numeric_split(0, 1, 1, 4.0, -7.5, -2.5)

        # Split the left leaf of the first tree in the first forest at X[,2] > 4.0
        forest2.add_numeric_split(0, 1, 1, 4.0, -1.5, -0.5)

        # Check that predictions are as expected
        pred1 = forest1.predict(forest_dataset)
        pred2 = forest2.predict(forest_dataset)
        pred_exp1 = np.array([-2.5, -7.5, -7.5, 5.0, 5.0, 5.0])
        pred_exp2 = np.array([-0.5, -1.5, 1.0, 1.0, 1.0, 1.0])

        # Assertion
        np.testing.assert_almost_equal(pred1, pred_exp1)
        np.testing.assert_almost_equal(pred2, pred_exp2)

        # Merge forests
        forest1.merge_forest(forest2)

        # Check that predictions are as expected
        pred = forest1.predict(forest_dataset)
        pred_exp = np.array([-3.0, -9.0, -6.5, 6.0, 6.0, 6.0])

        # Assertion
        np.testing.assert_almost_equal(pred, pred_exp)

        # Add constant to every value of the combined forest
        forest1.add_constant(0.5)

        # Check that predictions are as expected
        pred = forest1.predict(forest_dataset)
        pred_exp = np.array([7.0, 1.0, 3.5, 16.0, 16.0, 16.0])

        # Assertion
        np.testing.assert_almost_equal(pred, pred_exp)

        # Check that "old" forest is still intact
        pred = forest2.predict(forest_dataset)
        pred_exp = np.array([-0.5, -1.5, 1.0, 1.0, 1.0, 1.0])

        # Assertion
        np.testing.assert_almost_equal(pred, pred_exp)

        # Subtract constant back off of every value of the combined forest
        forest1.add_constant(-0.5)

        # Check that predictions are as expected
        pred = forest1.predict(forest_dataset)
        pred_exp = np.array([-3.0, -9.0, -6.5, 6.0, 6.0, 6.0])

        # Assertion
        np.testing.assert_almost_equal(pred, pred_exp)

        # Multiply every value of the combined forest by a constant
        forest1.multiply_constant(2.0)

        # Check that predictions are as expected
        pred = forest1.predict(forest_dataset)
        pred_exp = np.array([-6.0, -18.0, -13.0, 12.0, 12.0, 12.0])

        # Assertion
        np.testing.assert_almost_equal(pred, pred_exp)
